server/src/main/java/org/elasticsearch/cluster/routing/allocation/ShardsAvailabilityHealthIndicatorService.java

```diff
@@ -0,0 +1,191 @@
+        return DATA;
+    }
+
+    @Override
+    public HealthIndicatorResult calculate() {
+
+        var state = clusterService.state();
+        var shutdown = state.getMetadata().custom(NodesShutdownMetadata.TYPE, NodesShutdownMetadata.EMPTY);
+        var status = new ShardAllocationStatus();
+
+        for (IndexRoutingTable indexShardRouting : state.routingTable()) {
+            for (IndexShardRoutingTable shardRouting : indexShardRouting) {
+                status.addPrimary(shardRouting.primaryShard(), shutdown);
+                for (ShardRouting replicaShard : shardRouting.replicaShards()) {
+                    status.addReplica(replicaShard, shutdown);
+                }
+            }
+        }
+
+        return createIndicator(status.getStatus(), status.getSummary(), status.getDetails());
+    }
+
+    private static class ShardAllocationCounts {
+        private int unassigned = 0;
+        private int unassigned_restarting = 0;
+        private int initializing = 0;
+        private int started = 0;
+
+        public void increment(ShardRouting routing, NodesShutdownMetadata metadata) {
+            switch (routing.state()) {
+                case UNASSIGNED -> {
+                    if (isUnassignedDueToTimelyRestart(routing, metadata)) {
+                        unassigned_restarting++;
+                    } else {
+                        unassigned++;
+                    }
+                }
+                case INITIALIZING -> initializing++;
+                case STARTED -> started++;
+            }
+        }
+    }
+
+    private static boolean isUnassignedDueToTimelyRestart(ShardRouting routing, NodesShutdownMetadata metadata) {
+        var info = routing.unassignedInfo();
+        if (info == null || info.getReason() != UnassignedInfo.Reason.NODE_RESTARTING) {
+            return false;
+        }
+        var shutdown = metadata.getAllNodeMetadataMap().get(info.getLastAllocatedNodeId());
+        if (shutdown == null || shutdown.getType() != SingleNodeShutdownMetadata.Type.RESTART) {
+            return false;
+        }
+        var now = System.currentTimeMillis();
+        var restartingAllocationDelayExpiration = info.getUnassignedTimeInMillis() + shutdown.getAllocationDelay().getMillis();
+        return now <= restartingAllocationDelayExpiration;
+    }
+
+    private static class ShardAllocationStatus {
+        private final ShardAllocationCounts primaries = new ShardAllocationCounts();
+        private final ShardAllocationCounts replicas = new ShardAllocationCounts();
+
+        public void addPrimary(ShardRouting routing, NodesShutdownMetadata metadata) {
+            primaries.increment(routing, metadata);
+        }
+
+        public void addReplica(ShardRouting routing, NodesShutdownMetadata metadata) {
+            replicas.increment(routing, metadata);
+        }
+
+        public HealthStatus getStatus() {
+            if (primaries.unassigned > 0) {
+                return RED;
```

Comment:
I think we need to special handle new indices similar to [`ClusterShardHealth`](https://github.com/elastic/elasticsearch/blob/12ad399c488f0cc60e19b5e1b29c6d569cb4351a/server/src/main/java/org/elasticsearch/cluster/health/ClusterShardHealth.java#L201). I think we should report green for now for shards of such new indices in this PR, since there will need to be some interaction with allocation to see if this is unassigned due to throttling or because the shard cannot be assigned at all currently.

A separate count of the "new" unassigned/initializing shards seems necessary.

Reply:
Added